### PR TITLE
Make it possible to run Enso with assertions enabled

### PR DESCRIPTION
**Pull request by radeusgd**
[ci no changelog needed]

### Pull Request Description

Before, when running Enso with `-ea`, some assertions were broken and the interpreter would not start.

This PR fixes two very minor bugs that were the cause of this - now we can successfully run Enso with `-ea`, to test that any assertions in Truffle or in our own libraries are indeed satisfied.

Additionally, this PR adds a setting to SBT that ensures that IntelliJ uses the right language level (Java 17) for our projects.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- All code has been tested:
  - [ ] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH `./run dist` and `./run watch`.

### CODE REVIEW

**Review by jdunkerley — APPROVED** [review]

**Comment by radeusgd** [comment]
I also enabled assertions in the legacy workflow and will ask @mwu-tow to enable them in our new CI tests. Is everyone ok with that? cc: @jdunkerley @JaroslavTulach @hubertp @kustosz @ekmett

**Review by 4e6 — APPROVED** [review]
I'm ok with running tests with assertions enabled

**Review by hubertp — APPROVED** [review]
Maybe don't mix code formatting with the actual patch in the future.
Makes is hard to spot the actual lines that make the difference

**Comment by radeusgd** [comment]
> Maybe don't mix code formatting with the actual patch in the future. Makes is hard to spot the actual lines that make the difference

Sorry, it ran automatically for me and I thought that there isn't much sense in reverting it.

File: build.sbt [hunk 1]
```diff
@@ -1154,11 +1156,17 @@ val distributionEnvironmentOverrides = {
...
+    compilers.value,
+    javaVersion
+  ),
   // This dependency is needed only so that developers don't download Frgaal manually.
   // Sadly it cannot be placed under plugins either because meta dependencies are not easily
   // accessible from the non-meta build definition.
-  libraryDependencies +=  FrgaalJavaCompiler.frgaal
+  libraryDependencies += FrgaalJavaCompiler.frgaal,
+  // Ensure that our tooling uses the right Java version for checking the code.
+  javacOptions ++= Seq("-source", "17")
```
Comment (JaroslavTulach) [inline build.sbt]: That is not the right source level, right? Frgaal 18 uses `-source 18`.

File: build.sbt [hunk 2]
```diff
@@ -1212,7 +1224,9 @@ lazy val runtime = (project in file("engine/runtime"))
       s"--upgrade-module-path=${file("engine/runtime/build-cache/truffle-api.jar").absolutePath}"
     ),
     Test / fork := true,
-    Test / envVars ++= distributionEnvironmentOverrides ++ Map("ENSO_TEST_DISABLE_IR_CACHE" -> "false"),
+    Test / envVars ++= distributionEnvironmentOverrides ++ Map(
```
Comment (JaroslavTulach) [inline build.sbt:1230]: There is so many formatting changes! If they really have to be done, could they be performed in a separate commit (that doesn't need a review), please?

File: engine/runtime/src/main/scala/org/enso/compiler/codegen/IrToTruffle.scala [hunk 3]
```diff
@@ -1504,7 +1504,7 @@ class IrToTruffle(
               s"argument<${name.map(_.name).getOrElse(String.valueOf(position))}>"
 
             val section = value.location
-              .map(loc => source.createSection(loc.start, loc.end))
+              .map(loc => source.createSection(loc.start, loc.length))
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/scala/org/enso/compiler/codegen/IrToTruffle.scala:1507]: Ideally there should be a unit test that verifies this change.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/error/Warning.java [hunk 4]
```diff
@@ -54,7 +54,10 @@ boolean hasSourceLocation() {
     }
 
     @ExportMessage
-    SourceSection getSourceLocation() {
+    SourceSection getSourceLocation() throws UnsupportedMessageException {
+      if (location == null) {
+        throw UnsupportedMessageException.create();
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/error/Warning.java:59]: Can you explain why throwing `UnsupportedMessageException.create()` is the right choice? Of course, returning `null` isn't an option, but returning `Context.get(null).getBuiltins().nothing()` would be acceptable.

Ideally having a test somewhere to ensure this behavior ([stiffen the amoeba shape](http://wiki.apidesign.org/wiki/Amoeba#Minimizing_the_Amoeba_Effect)) would be beneficial.

File: .github/workflows/scala.yml [hunk 5]
```diff
@@ -255,6 +255,11 @@ jobs:
           gu install python
           gu install r
 
+      - name: Enable assertions for tests
+        shell: bash
+        run: |
+          echo "ENSO_JVM_OPTS=-ea" >> $GITHUB_ENV
```
Comment (JaroslavTulach) [inline .github/workflows/scala.yml]: How does that apply?

I know that when a `project-manager` is started, it pick this environment variable. But is that started? This `run` doesn't seem to contain any invocation...

**Review by JaroslavTulach — APPROVED** [review]
The set of necessary fixes is the best justification that enabling `-ea` in tests is needed!

File: engine/runtime/src/main/scala/org/enso/compiler/codegen/IrToTruffle.scala (hunk 3 above)
Comment (radeusgd) [reply engine/runtime/src/main/scala/org/enso/compiler/codegen/IrToTruffle.scala:1507]: Well, just running our tests with assertions enabled has found this, isn't it enough if we enable them then?

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/error/Warning.java (hunk 4 above)
Comment (radeusgd) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/error/Warning.java:59]: First of all `Context.get(null).getBuiltins().nothing()` has wrong type - the return type is `SourceSection` and that return type is required by the InteropLibrary.

But more importantly, this is required by the InteropLibrary contract (and checked there):
```
@throws UnsupportedMessageException if and only if {@link #hasSourceLocation(Object)} returns
     *             <code>false</code> for the same receiver.
```
the _if and only if_ is checked using assertions - if `hasSourceLocation` returns false, the assertions check that `getSourceLocation` will actually throw this exception.

As for tests - again, running the project with assertions enabled will trigger this, because just any call to `hasSourceLocation` with assertions enabled performs that check. So I'm not sure if it's worth duplicating that logic, if just any call is basically a test for that. And we do have such a call in our unit tests.

File: build.sbt (hunk 1 above)
Comment (hubertp) [reply build.sbt]: Speaking of which, we should probably have a global variable at the top of the build file so that this isn't c&p all over the place @radeusgd. I mean sth like `val JavaSourceLevel = ...`

File: .github/workflows/scala.yml (hunk 5 above)
Comment (radeusgd) [reply .github/workflows/scala.yml]: The environment is modified for all further steps.

BUT you are right! I always run the tests using the launcher, but the CI does not - it runs them directly, and that option is launcher-specific.

I will change it to `JAVA_TOOL_OPTIONS` and then it should be picked up by the JVM (I checked locally that it works on my machine).

File: build.sbt (hunk 2 above)
Comment (radeusgd) [reply build.sbt:1230]: Ok, I split the commit into two, the changes sans formatting can be seen here: https://github.com/enso-org/enso/pull/3450/files/0f942bb39be3ccfe33a4758b42a673f9e17a2ee8..17b867a1cec5f6d0ba8ca6d27e6e727b06fdf493

(just the files view with the first commit of the PR excluded)

File: build.sbt (hunk 1 above)
Comment (radeusgd) [reply build.sbt]: Yep.
